### PR TITLE
SeqLib integration (HTSlib-based BamTools switch)

### PR DESCRIPTION
**Pull request by walaj**
Hi Erik,

I have been working on an HTSlib based C++ API, as I was finding that with as much BAM access as I was doing, that switching to HTSlib was able to really speed things up over BamTools. I really liked the convenience of the BamTools API though, and worked to build something as simple for my own purposes ("SnowTools" library). I ended up cleaning it up, adding a ton of unit testing, and renamed it "SeqLib". I also got rid of c++11 dependence.

For a test of how it could perform (compute and as a simple API), I decided to switch out BamTools for SeqLib in Freebayes. The PR here has them merged, with a HAVE_BAMTOOLS preprocessor define that toggles them. I think every function used in BamTools has an equivalent one-line replacement in SeqLib. 

The performance gain was really quite dramatic:

```bash
## BamTools (original)
make CFLAGS='-DHAVE_BAMTOOLS=1'
time bin/freebayes --reference-fasta ref.fa test.bam > bamtools.vcf ## 20m49s
wc -l bamtools.vcf ## 17749
## SeqLib 
make clean && make
time bin/freebayes --reference-fasta ref.fa test.bam > seqlib.vcf ## 5m44s
wc -l seqlib.vcf ## 17749
## SeqLib with stdin
samtools view -h test.bam | bin/freebayes --reference-fasta ref.fa --stdin > seqlib_in.vcf
## SeqLib with native CRAM
bin/freebayes --reference-fasta ref.fa test.cram > seqlib_cram.vcf
```

Aside from what look like tiny roundoff differences in the QUAL field, a ``diff`` of the BamTools and SeqLib vcf files is nearly empty. (There is one last issue, in that for 3 of the 17k variants in my 1Gb test bam, the LEN field was different. They were all complex type. I think some minor one-line typo somewhere in how the LEN field is created for complex variants).

It also has multiple-bam reading functionality baked into the BamReader class. I haven't tested this in FreeBayes yet (I did implement though), and it works well in my own unit tests. 

Hopefully this solves the speed and the CRAM issues that have been raised with FreeBayes, without much effort on your part. It's good for me to have a nice example of the use of SeqLib outside of my own projects as well, and FreeBayes is a very visible and respected tool for this.

BTW you can also do error correction, BWA alignment, and sequence assembly with the SeqLib API, aside from just the BamTools replacement.

Thanks for considering, and I'm happy to answer questions / address any issues! Also, since it's such a big change, if you take the changes you might want to have them merge to a separate branch first.

Best,
Jeremiah

### CODE REVIEW

**Comment by jwalabroad** [comment]
e.g. On a 17k line VCF generated from a 1 gb BAM, after removing column 6 this is the only ``diff`` hit between bamtools and seqlib. Not sure where the different LEN is coming from... (seqlib has LEN=22, bamtools LEN=3)

```
10850c10850                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                
< 1     5731456 .       GAAAGTGGGGCAGATAAAAAATGG        GGA     .       AB=0.517483;ABP=3.38993;AC=1;AF=0.5;AN=2;AO=74;CIGAR=1M21D1M1X;DP=143;DPB=92.0417;DPRA=0;EPP=8.76177;EPPR=4.27115;GTI=0;LEN=22;MEANALT=7;MQM=36.9595;MQMR=60;NS=1;NUMALT=1;ODDS=148.496;PAIRED=0.959459;PAIREDR=1;PAO=16.5;PQA=438.5;PQR=859.5;PRO=32.5;QA=1499;QR=1714;RO=62;RPL=38;RPP=3.12768;RPPR=3.57068;RPR=36;RUN=1;SAF=38;SAP=3.12768;SAR=36;SRF=39;SRP=11.9764;SRR=23;TYPE=complex;technology.illumina=1       GT:DP:DPR:RO:QR:AO:QA:GL        0/1:143:143,74:62:1714:74:1499:-71.5815,0,-142.858                                                                                                                                                         
---                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        
> 1     5731456 .       GAAAGTGGGGCAGATAAAAAATGG        GGA     .       AB=0.517483;ABP=3.38993;AC=1;AF=0.5;AN=2;AO=74;CIGAR=1M21D1M1X;DP=143;DPB=92.0417;DPRA=0;EPP=8.76177;EPPR=4.27115;GTI=0;LEN=3;MEANALT=7;MQM=36.9595;MQMR=60;NS=1;NUMALT=1;ODDS=148.496;PAIRED=0.959459;PAIREDR=1;PAO=16.5;PQA=438.5;PQR=859.5;PRO=32.5;QA=1499;QR=1714;RO=62;RPL=38;RPP=3.12768;RPPR=3.57068;RPR=36;RUN=1;SAF=38;SAP=3.12768;SAR=36;SRF=39;SRP=11.9764;SRR=23;TYPE=complex;technology.illumina=1        GT:DP:DPR:RO:QR:AO:QA:GL        0/1:143:143,74:62:1714:74:1499:-71.5815,0,-142.858
```

Looks like multiple-bams is working now as well

**Comment by ekg** [comment]
@jwalabroad this is _amazing_ work. Thank you!

The performance improvement you note looks fantastic. I wonder if this might also help resolve memory problems we've had when loading many files simultaneously.

Moving to SeqLib looks like it will save a lot of researchers a lot of compute time.

I'll give the aggregate patch a read through to assure myself that there aren't any obvious issues that might have been hidden to you during implementation. Beyond that, I'll suggest a merge as soon as possible.

CC @zeeev @mlin for review assistance.

> e.g. On a 17k line VCF generated from a 1 gb BAM, after removing column 6 this is the only diff hit between bamtools and seqlib. Not sure where the different LEN is coming from... (seqlib has LEN=22, bamtools LEN=3)

The difference you point out is curious, and it looks like the bamtools one is wrong. I'm not totally sure what to make of it.

> BTW you can also do error correction, BWA alignment, and sequence assembly with the SeqLib API, aside from just the BamTools replacement.

I've been thinking about what it would take to enable assembly-based calling. It might be that we could exploit existing functions in the API. What do you think? Previously I was considering using [fermi-lite](https://github.com/lh3/fermi-lite).

**Comment by jwalabroad** [comment]
@ekg I'm very glad this is looking useful. BamTools was a life-saver for me going with C++ dev, but htslib is now just so fast I thought it was worth building the new API.

Regarding memory, I did a test of loading 5M 101-bp reads into memory using BamTools::BamAlignment or SeqLib::BamRecord. BamTools took 13.32 Gb and SeqLib 3.15 Gb. Hopefully the SeqLib move would alleviate a lot of memory overhead for joint calling.

I notice there are some issue reports here for things like NodeJS support, etc, from BamTools. Perhaps the right way to merge then would be to keep the preprocessor DHAVE_BAMTOOLS directive available, for anyone who needs some of the more auxiliary features that I haven't worked into SeqLib yet. I expect that for most users, a SeqLib default would be best though.

The assembly and error correction in SeqLib _is_ fermi-lite. I also separated out the error correction step (BFC) from the fermi-lite interface, so you can optionally error correct reads without the assembly. The assembly module in SeqLib right now works, but is the least tested of the functionality. It will likely take some exploring the parameter space of Fermi to really get what you would need, and there may be some extra SeqLib C++ API pieces needed to fully do that.

The BWA-MEM alignment, which I have been using to do contig-to-genome realignments and read-to-contig alignments on the fly are quite robust, and I've been using them routinely for ~1 year now. Might be useful for you in later versions of FreeBayes. I just opened up a repository for my own assembly-based SV/indel caller [Snowman][snow] which uses SeqLib, so feel free to try or review the code (eg the AlignedContig class) if you're thinking about assembly-based calling. For historical reasons, I use an integrated version of SGA as the assembler, but went with Fermi for SeqLib because of the MIT license and the much simpler code structure. 

[snow]: https://github.com/broadinstitute/SnowmanSV

**Comment by mlin** [comment]
This sounds pretty amazing! I don't have much hands-on familiarity with BamTools myself and wasn't aware of it being a critical performance bottleneck in relation to everything else freebayes does. @jwalabroad I was just wondering if you have any speculation on what makes it apparently so much slower than htslib as freebayes uses it? is it pervasive or might there be low-hanging fruit?

**Comment by jwalabroad** [comment]
It's a good question @mlin, and I'm not fully sure since I haven't taken a deep look into BamTools core operations. In SeqLib, the class that holds alignments (``BamRecord``) only has one data member, a shared pointer to an htslib ``bam1_t``, which stores data in a maximally compact form, allocating only enough bits as needed for each element and keeping them all together (the tradeoff is that modifying it takes some memory gymnastics). For instance, ``bam1_t`` stores all variable length data (name, cigar, tags, sequence and quality) in a single long ``uint8_t`` array. My guess is that BamTools is copying all of this data into a collection of separate ``std::string`` vars, even if its not required later. SeqLib builds it on the fly from ``bam1_t`` only when needed. 

Or maybe this isn't such a big deal and there is some other subtlety about inflating the bzipped BAM records that I'm missing. The memory difference is much easier to explain, as each ``std::string`` has constant overhead that a raw char array does not, and I believe BamTools also is storing different copies of the sequence (I think ``AlignedBases`` and ``QueryBases``). 

I don't think it's something weird in the FreeBayes implementation. My simple test of just reading in 5M reads and inflating the SeqLib sequence to a std::string still showed a >3x speed improvement

File: src/AlleleParser.cpp [hunk 1]
```diff
@@ -83,10 +84,43 @@ void AlleleParser::openBams(void) {
...
+        }
+    } else {
+      for (std::vector<std::string>::const_iterator i = parameters.bams.begin();
+           i != parameters.bams.end(); ++i) 
+        if (!bamMultiReader.Open(*i)) {
+	  ERROR("Could not open input BAM file: " + *i);
+	  exit(1);
+	}
+	else {
+	  /*if (!bamMultiReader.LocateIndexes()) {
```
Comment (ekg) [inline src/AlleleParser.cpp]: We could probably axe this commented-out code. Any reason to keep it?

**Comment by jwalabroad** [comment]
BamTools has a separate function for loading BAM indicies, whereas SeqLib finds them with htslib ``sam_index_load`` if needed during ``SetRegions()``. If no index is available, the BamTools version of FB will fail on LocateIndexes(), whereas the SeqLib version should emit a msg to stderr during ``SetRegions`` or ``SetMultipleRegions`` and return ``false``. For consistency, it should probably have a check that ``SetRegions`` returns true, otherwise stop FB.

**Comment by zeeev** [comment]
I've started testing the code.  It looks like bamleftalign will need to be updated as well.

**Comment by walaj** [comment]
@zeeev Ahh yes, I forgot to test bamleftalign. Just committed changes to bamleftalign.cpp to use the macros from LeftAlign.h. Also added in compress/uncompress output option for SeqLib, as in the original verson. I don't think there should be any difference between the BT/SeqLib versions now.

```
make CFLAG='-DHAVE_BAMTOOLS=1'
time samtools view -h $na12878 1:1,000,000-10,000,000 | bin/bamleftalign > bamtools.leftalign.sam #1m40s
make clean && make
time samtools view -h $na12878 1:1,000,000-10,000,000 | bin/bamleftalign > seqlib.leftalign.sam # 1m19s
diff bamtools.leftalign.sam seqlib.leftalign.sam ## empty
```

**Comment by ekg** [comment]
Before this goes stale, I think it's time to pull it in.

@walaj the only concern I have is that SeqLib has a very heavy git repository that can take a long time to pull down. Could we consider ways to improve this? I'll merge without a solution, but I think we should address this.

**Comment by walaj** [comment]
Good idea re: cleaning the SeqLib repos. Looks like I had been storing some static libraries for some time, which is making this so heavy. Cleaning now, will push a lighter version shortly I suspect.

**Comment by walaj** [comment]
Reduced from 123Mb to 37Mb. There's still a pretty large history, but this got rid of all of the low hanging fruit. Not sure how to reduce this further without a lot of git tricks, but open to suggestions.

**Comment by ekg** [comment]
@walaj sorry, it looks like my merge of another PR broke the potential merge.

**Comment by walaj** [comment]
@ekg @sambrightman 

I thought I had merged this correctly, but it looks like I'm failing the new region tests, not sure why though. Can either of you make sense of the failed tests from the travis log? (example below)

```
t/00_region_and_target_handling.t .. 
1..11
[samopen] SAM header is present: 1 sequences.
 HAVE SEQLIB 
ok 1 - ref:4-5 ref:6-7 are empty
 HAVE SEQLIB 
ok 2 - ref:4 ref:6 are empty
 HAVE SEQLIB 
not ok 3 - ref:5-6 ref:7-8
#   Failed test 'ref:5-6 ref:7-8'
#   at t/00_region_and_target_handling.t line 101.
 HAVE SEQLIB 
not ok 4 - ref:5 ref:7
#   Failed test 'ref:5 ref:7'
#   at t/00_region_and_target_handling.t line 104.
```

**Comment by sambrightman** [comment]
Yes, it looks like your `HAVE SEQLIB` debugging is being captured and breaking the comparison. Remove the debugging?

I can't build on Mac with this branch by the way: lots of errors related to `tr1`.

**Comment by walaj** [comment]
@sambrightman thanks for the info, I'll remove the print statement and try again

The tr1 errors are because I had a pre-processor directive that would use tr1 if anything less than c++11 was being used, since shared_ptr is defined in tr1 for pre c++11. I guess this is problematic on a Mac. I changed the default so that if on mac (__APPLE__ is defined), SeqLib will used the std:: functions instead of tr1::  It compiles now on OSX.

@ekg Compiling now on clang on OSX per @sambrightman request, I am seeing a lot of compiler warnings. Most are innocuous, but I think some are small bugs (eg Genotype.cpp at 1306 is an unused comparison, I think it should just be a single '='). I may as well get rid of them while I'm at it if that's OK. Mostly just putting parentheses around multiple comparisons.

**Comment by sambrightman** [comment]
I also noticed those warnings, especially the double equals - but didn't have time to investigate.

**Comment by walaj** [comment]
@ekg I fixed the Genotype.cpp line 1306 bug, and added a few parentheses around elsewhere to stop warnings. Also two functions in AlleleParser were "bool" type with no return, so I set them to void.

@sambrightman I am still failing the new region tests, even after removing the debug print statement. Can you take a look and see what might be going on? I'm having a hard time making sense of how this test unit works.

**Comment by sambrightman** [comment]
@walaj if I pull your current master and update submodules, compilation still fails on Mac. The commit history could do with some rebasing - besides all the change/revert stuff, in some cases the messages don't actually describe the content at all.

It's not just the region tests that are failing (which I didn't write). It looks like there is no output produced for the failing ones - maybe look into that? e.g. take the simple SAM from the test and see if you can get a call.

You can reproduce them by directly copy & pasting from the script if you don't understand something:

```
# setup

bind 'set disable-completion on' # we want to paste some tabs
PATH=~/code/samtools:$PATH
root=.
ref=/tmp/test.fasta
bam=/tmp/test.bam

# direct copy & paste from tests

cat >$ref <<REF
>ref
ATCGGCTAAAA
REF
samtools faidx $ref

samtools view -S -b - >$bam <<SAM
@HD	VN:1.5	SO:coordinate
@SQ	SN:ref	LN:11
alt	0	ref	1	30	1X1=2X1=1X1=1X2=1X	*	0	0	GTTAGGTTAAC	*
SAM
samtools index $bam

function run_freebayes() {
    ($root/bin/freebayes "$@" \
        --haplotype-length 0 --min-alternate-count 1 \
        --min-alternate-fraction 0 --pooled-continuous --report-monomorphic \
        --ploidy 1 \
        -f $ref $bam \
        | grep -vE "^#" | cut -f1-5)
}

run_freebayes
```

**Comment by walaj** [comment]
@sambrightman Thanks for the info, this was really helpful. Indeed there was an insidious bug that only came up in the unit test. X and = cigar types were not being counted as aligned bases in one seqlib function, which caused freebayes to skip the alignment in the test BAM. My set of test BAMs didn't have X or = cigar type, so I didn't notice until now. Unit tests are terribly underrated in bioinformatics, and stuff like this really makes me appreciate them...

The OSX issue: the SeqLib repos was fixed, but the freebayes pointer to SeqLib was still at the older commit. Updated now, and confirmed that freebayes compiles on OSX.

The official freebayes repos and my PR now both pass the same tests and are passing on TravisCI.

@ekg I believe this addresses the outstanding issues. Let me know if you need anything for this PR

**Comment by sambrightman** [comment]
@walaj I agree. Really great to see that they already caught an issue in this case - perhaps you could help improve matters by ensuring that SeqLib is well-tested?

**Comment by walaj** [comment]
@sambrightman I would argue that it is already very well tested. I have unit tests covering ~87% of the code. SeqLib has also been used now in production on thousands of whole genomes. Here, the problem was in an edge case for a single function that serves only as a direct replacement for ``AlignedBases.size()``, which I made just for this PR. I had to sort of reverse engineer what ``AlignedBases`` was doing (e.g. counting both D and I as query aligned bases, which is different than how htslib treats them). I think the unit tests did exactly what they were meant to here, and both FreeBayes and SeqLib seem as tested as one could reasonably expect.

Not sure why the FreeBayes region and target handling tests weren't run before the merges with your PR, but everything looks fine now!

**Comment by sambrightman** [comment]
> Not sure why the FreeBayes region and target handling tests weren't run before the merges with your PR, but everything looks fine now!

They were just in the wrong directory.

**Comment by ekg** [comment]
@walaj amazing work. Thank you many times over for this contribution!

@sambrightman thank you as well for following up on this PR.

**Comment by walaj** [comment]
Great! Thank you Erik and Sam for all the help as well.
